src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -7420,7 +7420,38 @@ public StatementSyntax ParseStatement()
         }
 
         private StatementSyntax ParsePossiblyAttributedStatement()
-            => ParseStatementCore(ParseAttributeDeclarations(inExpressionContext: true), isGlobal: false);
+            => ParseStatementCore(ParseStatementAttributeDeclarations(), isGlobal: false);
+
+        private SyntaxList<AttributeListSyntax> ParseStatementAttributeDeclarations()
+        {
+            if (this.CurrentToken.Kind != SyntaxKind.OpenBracketToken)
+                return default;
+
+            // See if we should treat this as a collection expression.  At the top-level or statement-level, this should
+            // only be considered a collection if followed by a `.`, `?` or `!` (indicating it's a value, not an
+            // attribute).
+            var resetPoint = GetResetPoint();
+
+            // Grab the first part as a collection expression.
+            ParseCollectionCreationExpression();
+
+            // Continue consuming element access expressions for `[x][y]...`.  We have to determine if this is a
+            // collection expression being indexed into, or if it's a sequence of attributes.
+            while (this.CurrentToken.Kind == SyntaxKind.OpenBracketToken)
+                ParseBracketedArgumentList();
+
+            // Check the next token to see if it indicates the `[...]` sequence we have is a term or not. 
+            var isCollectionExpression = this.CurrentToken.Kind is SyntaxKind.DotToken or SyntaxKind.QuestionToken or SyntaxKind.ExclamationToken;
```

Comment:
It feels like this should be scanning for either a member modifier or something that could be a type: if it finds one of those, it must be an attribute. Otherwise, it must be a collection initializer. Consider `[] + new List<int>()`, or `[] == new List<int>()`, or number of other possible scenarios like them. A hardcoded list of allowable tokens feels too fragile.

Reply:
so those don't work here fro the same reason that:

`a + b;` is not a legal statement.  Will discuss more offline.

Reply:
Further offline discussion has revealed missing cases such as `[() => {}][0]();` or `[0]++;`. Cyrus will file a tracking issue.